train_decoder.py

```diff
@@ -0,0 +1,488 @@
+                #     average_loss += loss.item()
+                with autocast(enabled=amp):
+                    for unet in range(1, len(decoder.unets)+1):
+                        loss = decoder.forward(img.float(), image_embed=emb.float(), unet_number=unet)
+                        average_loss += loss.item()
+
+                if i % 10 == 0:
+                    print(f"Epoch {epoch}/{epochs} - {sample / (time.time() - start_time):.2f} samples/sec")
+                    print(f"Loss: {average_loss / (i+1)}")
+                    print("")
+
+                if validation_length is not None and i >= validation_length:
+                    break
+            average_loss /= sample
+            wandb.log({
+                "Validation loss": average_loss,
+                "Epoch": epoch,
+                "Sample": sample
+            }, step=step)
+
+        print(f"=========== Starting Sampling {epoch} ===========")
+        if n_samples is not None and n_samples > 0:
+            # Then we will select the top N samples from test and use trainer.sample(image_embed) to generate samples
+            test_images = generate_samples(decoder, dataloaders["test"], epoch, device, step, n=n_samples, text_prepend="Test: ")
+            train_images = generate_samples(decoder, dataloaders["train"], epoch, device, step, n=n_samples, text_prepend="Train: ")
+            wandb.log({
+                "Test samples": test_images,
+                "Train samples": train_images
+            }, step=step)
+
+        print(f"=========== Starting Saving {epoch} ===========")
+        save_trainer(save_path, trainer, epoch, step)
+
+    
+def main(
+    webdataset_base_url,
+    embeddings_url,
+    device_name="cuda",
+    cuda=True,
+    epochs=20,
+    epoch_length=None,
+    validation_length=None,
+    n_samples=5,
+    do_overfit=False,
+    save_path="./models",
+    learning_rate=1.2e-4,
+    weight_decay=0,
+    batch_size=32,
+    img_size=(64, 64),
+    use_ema=True,
+    start_shard = 169230,
+    end_shard = 169400,
+):
+    # Create the save path
+    if not os.path.exists(save_path):
+        os.makedirs(save_path)
+    cuda = cuda and torch.cuda.is_available()
+    if cuda:
+        device = torch.device(device_name)
+        torch.cuda.set_device(device)
+    else:
+        device = torch.device("cpu")
+    all_shards = list(range(start_shard, end_shard+1))
+
+    dataloaders = create_dataloaders (
+        available_shards=all_shards,
+        webdataset_base_url=webdataset_base_url,
+        embeddings_url=embeddings_url,
+        num_workers=4,
+        batch_size=batch_size,
+        shuffle_train=False,
+        shuffle_val_test=False,
+        img_size=img_size,
+        img_preproc = None,
+        index_width=4,
+        train_prop = 0.75,
+        val_prop = 0.15,
+        test_prop = 0.10
+    )
+
+    decoder, (unet_config, decoder_config) = create_decoder(device)
+
+    wandb.init(
+        entity="Veldrovive",
```

Comment:
^ Perhaps get this from environment or somewhere else, so others can use your trainer?

Reply:
Yes, this is just to get some feedback on if the training is set up correctly. CLI args are the next step once it produces any results on a larger dataset.

Reply:
The option to use any entity has now been added along with other configuration options.